### DISCUSSION TITLE
Exceute tool after/from generate_oai_reply

### DISCUSSION THREAD

**Discussion by WebsheetPlugin** [microsoft/autogen#2218] [Q&A]
**My code:**
def` user_msg(msg_text):
    return {"role": "user", "content": msg_text}

llm_config={"config_list": config_list}

analyst = AssistantAgent(
    name="Analyst",
    system_message="""Your an analyst.  You do XYZ.""",
    llm_config=llm_config,
)

report = """Multiply 110 * 73282 and 90 * 9292"""

analyst.register_for_llm(name="calculator", description="A simple calculator")(calculator)
print(analyst.llm_config["tools"]) # debug tools

analysis =  analyst.generate_oai_reply([user_msg(report)])
print(f"Analysis: {analysis}")

**Prints:**
Analysis: (True, {'content': None, 'role': 'assistant', 'function_call': None, 'tool_calls': [{'id': 'call_RPdBCRwuUOVA4t5WpyNO8Kpr', 'function': {'arguments': '{"a": 110, "b": 73282, "operator": "*"}', 'name': 'calculator'}, 'type': 'function'}, {'id': 'call_zTg9NHo25Q7OnqMXEed0jzpo', 'function': {'arguments': '{"a": 90, "b": 9292, "operator": "*"}', 'name': 'calculator'}, 'type': 'function'}]})

**I was able to use generate_oai_reply to mimic the openai tool call feature. Is there a way to use the analysts' response to execute those tools now? Or do I have to do it on my own? I don't plan to use user_proxy or agent chats**

**Comment by sonichi** [comment]
You can use `register_nested_chats` to register an inner chat between an AssistantAgent and a UserProxyAgent which will perform the tool call + tool execution in the chat.
cc @qingyun-wu 
I think this is a good example to add in the notebooks.

**Reply by ekzhu** [reply]
The chess example notebook already does this: https://microsoft.github.io/autogen/docs/notebooks/agentchat_nested_chats_chess

**Comment by ekzhu** [comment]
Check out this tutorial section https://microsoft.github.io/autogen/docs/tutorial/tool-use#how-to-hide-tool-usage-and-code-execution-within-a-single-agent
